Add (and use) COND

notmgsk opened the pull request.

notmgsk commented on src/library.lisp:
```diff
@@ -37,6 +37,15 @@
+                      ,then-a
+                      (coalton::cond ,@clauses))
+         (cl:progn
+           (cl:assert (cl:eql cl:t clause-a) () "COND must have a catch-all T clause.")
```
`cl:t` or `otherwise`?

notmgsk replied:
or `else`?

notmgsk committed:
Add COND

Rewrite GCD to use COND

Use 'else' rather than 'cl:t'

stylewarning commented on src/library.lisp:
```diff
@@ -37,6 +37,15 @@
      (coalton:true ,then)
      (coalton:false ,else)))
 
+(cl:defmacro coalton::cond ((clause-a then-a) cl:&rest clauses)
```
```suggestion
(cl:defmacro coalton:cond ((clause-a then-a) cl:&rest clauses)
```

stylewarning commented on src/library.lisp:
```diff
@@ -37,6 +37,15 @@
      (coalton:false ,else)))
 
+(cl:defmacro coalton::cond ((clause-a then-a) cl:&rest clauses)
+  (cl:if (cl:not (cl:null clauses))
```
```suggestion
  (cl:if (cl:not (cl:endp clauses))
```

stylewarning commented on src/library.lisp:
```diff
@@ -37,6 +37,15 @@
+  (cl:if (cl:not (cl:null clauses))
+         `(coalton:if ,clause-a
+                      ,then-a
+                      (coalton::cond ,@clauses))
```
```suggestion
                      (coalton:cond ,@clauses))
```

stylewarning commented on src/library.lisp:
```diff
@@ -37,6 +37,15 @@
+                      ,then-a
+                      (coalton::cond ,@clauses))
+         (cl:progn
+           (cl:assert (cl:eql 'else clause-a) () "COND must have a catch-all T clause.")
```
```suggestion
           (cl:assert (cl:eq 'coalton:else clause-a) () "COND must have an ELSE clause.")
```

stylewarning commented on src/package.lisp:
```diff
@@ -26,6 +26,7 @@
    #:let
    #:letrec
    #:if
+   #:cond
```
```suggestion
   #:cond
   #:else          ; syntax for COND
```

stylewarning commented on src/library.lisp:
```diff
@@ -132,21 +141,15 @@
-            v
-            (if (zerop v)
-                u
-                (if (evenp u)
-                    (if (evenp v)
-                        (double (gcd (half u) (half v)))
-                        (gcd (half u) v))
-                    (if (evenp v)
-                        (gcd u (half v))
-                        (if (> u v)
-                            (gcd (half (- u v)) v)
-                            (gcd (half (- v u)) u)))))))))
+    (cond ((= u v)   u)
+          ((zerop u) v)
+          ((zerop v) u)
+          ((evenp u) (if (evenp v)
+                         (double (gcd (half u) (half v)))
+                         (gcd (half u) v)))
+          (else      (cond ((evenp v) (gcd u (half v)))
+                           ((> u v)   (gcd (half (- u v)) v))
+                           (else      (gcd (half (- v u)) u)))))))
```
Beautiful!

notmgsk committed:
colons

notmgsk committed:
Review comments

notmgsk committed:
Remove some coalton: prefixes

notmgsk committed:
Remove some coalton: prefixes, and Capitalise True/False

notmgsk commented:
@stylewarning I added a couple more changes that might be controversial. Namely, I removed the `coalton:` prefixes in places where they're not technically required, and I capitalised the symbols `false` and `true`. I can revert those if you think it's better to be clear about where those symbols are coming from.

stylewarning commented on src/library.lisp:
```diff
@@ -34,12 +34,21 @@
 
 (cl:defmacro coalton:if (expr then else)
   `(match ,expr
-     (coalton:true ,then)
```
Removal here is OK. I mostly want to keep the prefixes at definitions.

stylewarning commented on src/library.lisp:
```diff
@@ -34,12 +34,21 @@
 (cl:declaim (cl:inline lisp-boolean-to-coalton-boolean))
 (cl:defun lisp-boolean-to-coalton-boolean (x)
-  (cl:if x coalton:true coalton:false))
+  (cl:if x True False))
```
good change, I prefer this
